File: LeetCode/1-100/14. Longest Common Prefix/Solution.java
class Solution {
    public String longestCommonPrefix(String[] strs) {
        if(strs == null || strs.length == 0) 
            return "";
        if(strs.length == 1) 
            return strs[0];
        String result = commonPrefix(strs[0], strs[1]);
        for(int i = 2; i < strs.length; i++){
            result = commonPrefix(result, strs[i]);
        }
        return result;
    }
    private String commonPrefix(String s1, String s2){
        if(s1 == null || s2 == null) 
            return "";
        int len = s1.length();
        for(int i = len; i >= 0; i--){
            String prefix = s1.substring(0, i);
            if(s2.startsWith(prefix)) 
                return prefix;
        }
        return "";
    }
}

File: LeetCode/1-100/14. Longest Common Prefix/Solution.cpp
/*
     Time Complexity: O(N)  // We go through every charcter in every string
     Space Complexity: O(1)
*/

class Solution {
public:
  string longestCommonPrefix(vector<string> &strs) {
    string res = "";
    for (int i = 0; i < strs[0].length(); i++) {
      for (auto s : strs) {
        if (i == s.length() || s[i] != strs[0][i])
          return res;
      }
      res += strs[0][i];
    }
    return res;
  }
};

// Another approach
class Solution {
public:
  string longestCommonPrefix(vector<string> &strs) {
    string common = strs[0];
    int minLen = strs[0].length();
    for (int i = 0; i < strs.size(); i++) {
      if (strs[i].length() < minLen) {
        minLen = strs[i].length();
      }
    }
    int i = 0;
    while (i < minLen) {
      char currCharacter = strs[0][i];
      for (int j = 1; j < strs.size(); j++) {
        if (currCharacter == strs[j][i])
          continue;
        else
          return common.substr(0, i);
      }
      i++;
      common += strs[0][i];
    }

    return common.substr(0, i);
  }
};